[17.0][MIG] web_form_banner: Migration to 17.0

AungKoKoLin1997 opened the pull request:
Based on https://github.com/OCA/web/pull/3298

The module adds configurable banners for backend form views. Define rules per model
(and optionally per view) to show context-aware alerts with a chosen severity (info/warning/danger).

Messages can be plain text with ${placeholders} or fully custom HTML; visibility,
severity, and values are computed server-side via a safe Python expression.

Banners are injected just before or after a target node (default: //sheet) and refresh
on form load/save/reload.

@qrtl QT5747

yostashiro committed:
[ADD] web_form_banner

AungKoKoLin1997 committed:
[IMP] web_form_banner: black, isort, prettier

AungKoKoLin1997 committed:
[MIG] web_form_banner: Migration to 15.0

Refactor JS:
- Switch to the ESM style
- Change var to const/let
- Remove unused parts (diff, el.dataset.wfbTriggerFields)

Miscellaneous improvements:
- list view optional
- Switch archive box icon to banner

Co-authored-by: Yoshi Tashiro <tashiro@quartile.co>

AungKoKoLin1997 committed:
[IMP] web_form_banner: pre-commit auto fixes

AungKoKoLin1997 committed:
[MIG] web_form_banner: Migration to 16.0

yostashiro committed:
[IMP] web_form_banner: Partially support in-group target XPath

- Add (partial) support for elements inside groups to be a target
  XPath without distorting the presentation
- Remove obsolete bits

AungKoKoLin1997 committed:
[IMP] web_form_banner: pre-commit auto fixes

AungKoKoLin1997 committed:
[MIG] web_form_banner: Migration to 17.0

yostashiro commented on web_form_banner/models/web_form_banner_rule.py:
```diff
@@ -39,7 +38,7 @@ def _extract_m2o_id(v):
     if isinstance(v, int):
         return v
-    if isinstance(v, (list, tuple)) and v and isinstance(v[0], int):
+    if isinstance(v, list | tuple) and v and isinstance(v[0], int):
```
I don't quite understand this change. Why do we need to do this?

AungKoKoLin1997 replied:
This is the new style, introduced in PEP 604 (Python 3.10+) and style guide (Ruff) forces us to use that style in the pre-commit.

yostashiro commented on web_form_banner/models/web_form_banner_rule.py:
```diff
@@ -172,7 +171,6 @@ def _build_form_url(self, rec):
 
-    @lru_cache(maxsize=1)
     def _base_eval_ctx_static(self):
         # Only static, import-heavy items
```
Please add @api.model. We may perhaps remove the the comment as it doesn't say much more than the method name.

I suppose the `lru_cache` removal should also be applied to the source versions?

yostashiro commented on web_form_banner/static/src/js/web_form_banner.esm.js:
```diff
@@ -26,16 +27,20 @@ const safe = async (fn, fb) => {
 };
 
 function normalizeValue(v) {
```
Please avoid removing comments unless they are just unnecessary noise. Also avoid immaterial style changes.

yostashiro commented:
Dynamic banner update seems a bit slower than before. Is there room for improvement?

AungKoKoLin1997 commented:
> Dynamic banner update seems a bit slower than before. Is there room for improvement?

@yostashiro I compared between v16 and v17 in runboat with the latest changes and it is like the same.

AungKoKoLin1997 committed:
[IMP] web_form_banner: imp readme

yostashiro reviewed (approved):
Code review and functional test. LGTM.

yostashiro reviewed (changes requested):
There is an issue with banner reload with previous/next button. See https://github.com/OCA/web/pull/3308#pullrequestreview-3329902995.

AungKoKoLin1997 committed:
[FIX] web_form_banner

- Fix not updating banner when pager is update or duplicate or create
- Fix presentation of help in web form banner rule

yostashiro commented on web_form_banner/views/web_form_banner_rule_views.xml:
```diff
@@ -0,0 +1,202 @@
+                                    Should be used instead of <code
+                                    >record</code> when your rule is triggered dynamically by an
+                                    update to a trigger field. It doesn't include any values from complex fields
+                                    (x2many/reference, etc).
```
Minor adjustment.
```suggestion
                                    (one2many/reference, etc).
```

yostashiro reviewed (approved):
Code review and functional test. LGTM.

nobuQuartile reviewed (approved):
LGTM
Functional test

AungKoKoLin1997 committed:
[IMP] web_form_banner: add summary and maintainers to manifest
